[Hydrawise] Binding stops polling after error / exception

Openhab : 3.1.0
Hydrawaise Binding : 3.1.0.202106121717

I've noticed a few times where the binding seems to stop updating after a few days and stale data is left displayed on my sitemaps.

I ran the binding in debug and caught the exception, which seems to be related to the account (possibly bad password / not available) and whilst this update might be dropped, the binding should re-attempt a minute later, then a minute after that, etc:

```
2022-06-20 19:22:23.638 [TRACE] [l.api.graphql.HydrawiseGraphQLClient] - Received Response: {"data":{"me":{"email":"xxxxxx@gmail.com","lastContact":"Sat, 21 May 22 04:17:32 +0000","controllers":[{"id":xxxxxxxx,"name":"My Controller","status":{"summary":"All
 good!","online":true,"lastContact":{"timestamp":1655669237}},"location":{"coordinates":{"latitude":-33.7697072,"longitude":151.2819771},"forecast":[{"time":"Mon, 20 Jun 22 00:00:00 +1000","updateTime":"Mon, 20 Jun 22 05:14:27 +1000","conditions":"Scattered Show
ers","averageWindSpeed":{"value":12,"unit":"km\/h"},"highTemperature":{"value":18,"unit":"\u00b0C"},"lowTemperature":{"value":9,"unit":"\u00b0C"},"probabilityOfPrecipitation":45.44,"precipitation":{"value":0.03,"unit":"mm"},"averageHumidity":69},{"time":"Tue, 21
 Jun 22 00:00:00 +1000","updateTime":"Mon, 20 Jun 22 05:14:27 +1000","conditions":"Light Rain","averageWindSpeed":{"value":16,"unit":"km\/h"},"highTemperature":{"value":20,"unit":"\u00b0C"},"lowTemperature":{"value":10,"unit":"\u00b0C"},"probabilityOfPrecipitation":77.2,"precipitation":{"value":0.92,"unit":"mm"},"averageHumidity":64},{"time":"Wed, 22 Jun 22 00:00:00 +1000","updateTime":"Mon, 20 Jun 22 05:14:27 +1000","conditions":"Sunny","averageWindSpeed":{"value":20,"unit":"km\/h"},"highTemperature":{"value":18,"unit":"\u00b0C"},"lowTemperature":{"value":8,"unit":"\u00b0C"},"probabilityOfPrecipitation":19.04,"precipitation":{"value":0,"unit":"mm"},"averageHumidity":55}]},"zones":[{"id":4473267,"name":"Bamboo","status":{"suspendedUntil":null},"icon":{"id":1,"fileName":"leaf.png","customImage":null},"number":{"value":1,"label":"Zone 1"},"scheduledRuns":{"summary":"Bamboo will next run on Tuesday, 21st of June 2022, 07:00AM (Sydney Time) for 20 minutes. Normal watering cycle","currentRun":null,"nextRun":{"id":"21412093300","startTime":{"timestamp":1655758800},"endTime":{"timestamp":1655760000},"duration":20}}},{"id":4469122,"name":"Trees and Shrubs","status":{"suspendedUntil":null},"icon":{"id":4,"fileName":"flower.png","customImage":null},"number":{"value":3,"label":"Zone 3"},"scheduledRuns":{"summary":"Trees and Shrubs will next run on Wednesday, 22nd of June 2022, 07:00AM (Sydney Time) for 15 minutes. Normal watering cycle","currentRun":null,"nextRun":{"id":"21412093301","startTime":{"timestamp":1655845200},"endTime":{"timestamp":1655846100},"duration":15}}},{"id":4472303,"name":"Grass Pool","status":{"suspendedUntil":null},"icon":{"id":10,"fileName":"grass.png","customImage":null},"number":{"value":4,"label":"Zone 4"},"scheduledRuns":{"summary":"Grass Pool will next run on Wednesday, 22nd of June 2022, 07:15AM (Sydney Time) for 20 minutes. Normal watering cycle","currentRun":null,"nextRun":{"id":"21412093302","startTime":{"timestamp":1655846100},"endTime":{"timestamp":1655847300},"duration":20}}},{"id":4472467,"name":"Grass Beds","status":{"suspendedUntil":null},"icon":{"id":10,"fileName":"grass.png","customImage":null},"number":{"value":5,"label":"Zone 5"},"scheduledRuns":{"summary":"Grass Beds will next run on Wednesday, 22nd of June 2022, 07:35AM (Sydney Time) for 20 minutes. Normal watering cycle","currentRun":null,"nextRun":{"id":"21412093303","startTime":{"timestamp":1655847300},"endTime":{"timestamp":1655848500},"duration":20}}}],"sensors":[]}]}}}
2022-06-20 19:22:23.639 [TRACE] [l.handler.HydrawiseControllerHandler] - onData my controller id xxxxxxxxx
2022-06-20 19:22:23.639 [TRACE] [l.handler.HydrawiseControllerHandler] - Updating Controller xxxxxxxx sensors [] forecast [org.openhab.binding.hydrawise.internal.api.graphql.schema.Forecast@7092d538, org.openhab.binding.hydrawise.internal.api.graphql.schema.Forecast@4caed024, org.openhab.binding.hydrawise.internal.api.graphql.schema.Forecast@4caa9a02] 
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast1 temperaturehigh ?C 18
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast1 temperaturelow ?C 9
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast2 temperaturehigh ?C 20
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast2 temperaturelow ?C 10
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast3 temperaturehigh ?C 18
2022-06-20 19:22:23.640 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast3 temperaturelow ?C 8
2022-06-20 19:22:23.640 [TRACE] [l.handler.HydrawiseControllerHandler] - Updateing Zone zone1 Bamboo 
2022-06-20 19:22:23.640 [TRACE] [l.handler.HydrawiseControllerHandler] - Updateing Zone zone2 Trees and Shrubs 
2022-06-20 19:22:23.640 [TRACE] [l.handler.HydrawiseControllerHandler] - Updateing Zone zone3 Grass Pool 
2022-06-20 19:22:23.640 [TRACE] [l.handler.HydrawiseControllerHandler] - Updateing Zone zone4 Grass Beds 
2022-06-20 19:23:23.641 [TRACE] [l.api.graphql.HydrawiseGraphQLClient] - Sending Request: {"query":"{\n  me {\n    email\n    lastContact\n      controllers {\n      id\n      name\n        status {\n        summary\n        online\n        lastContact {\n          timestamp\n        }\n      }    \n      location {\n        coordinates {\n          latitude\n          longitude\n        }\n        forecast(days: 3) {\n          time\n          updateTime\n          conditions\n          averageWindSpeed {\n            value\n            unit\n          }\n          highTemperature {\n            value\n            unit\n          }\n          lowTemperature {\n            value\n            unit\n          }\n          probabilityOfPrecipitation\n          precipitation {\n            value\n            unit\n          }\n          averageHumidity\n        }\n      }\n      zones {\n        id\n        name\n        status {\n          suspendedUntil {\n            timestamp\n          }     \n        }\n        icon {\n          id\n          fileName\n          customImage {\n            id\n            url\n          }\n        }\n        number {\n          value\n          label\n        }\n        scheduledRuns {\n          summary\n          currentRun{\n             id\n            startTime {\n              timestamp\n            }\n            endTime {\n              timestamp\n            }\n            duration\n            status {\n              value\n              label\n            }\n          }\n          nextRun {\n            id\n            startTime {\n              timestamp\n            }\n            endTime {\n              timestamp\n            }\n            duration\n          }\n        }\n      }\n      sensors {\n        id\n        name\n        input {\n          number\n          label\n        }\n        status {\n          active\n          waterFlow {\n            value\n            unit\n          }\n        }\n        model {\n          modeType\n          active\n          offLevel\n          offTimer\n          delay\n        }\n      }\n    }\n  }\n}"}
**2022-06-20 19:23:24.364 [TRACE] [l.api.graphql.HydrawiseGraphQLClient] - Received Response: 
2022-06-20 19:23:24.364 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
java.lang.NullPointerException: null
        at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:192) ~[?:?]
        at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:183) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]**
2022-06-20 19:30:00.928 [INFO ] [enhab.core.model.script.garage.rules] - Garage Door is Closed
2022-06-20 19:45:00.564 [INFO ] [enhab.core.model.script.lights.rules] - Bedtime, some lights off, some low
```

Sorry for the late reply, it took me some time to catch a relevant error message.
Is this of any help for you?
Kind regards, Joachim

```
2022-07-21 18:11:19.520 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast2 temperaturelow °C 16

2022-07-21 18:11:19.526 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast3 temperaturehigh °C 26

2022-07-21 18:11:19.529 [DEBUG] [l.handler.HydrawiseControllerHandler] - TEMP forecast3 temperaturelow °C 15

2022-07-21 18:29:55.722 [DEBUG] [rnal.handler.HydrawiseAccountHandler] - Retrying failed poll

org.openhab.binding.hydrawise.internal.api.HydrawiseConnectionException: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@1066aff::DecryptedEndPoint@df1913{l=/192.168.178.xxx:34194,r=app.hydrawise.com/44.236.19.36:443,OPEN,fill=-,flush=-,to=1116380/0}

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:318) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLQuery(HydrawiseGraphQLClient.java:256) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.queryControllers(HydrawiseGraphQLClient.java:117) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:188) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:183) ~[?:?]

Caused by: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@1066aff::DecryptedEndPoint@df1913{l=/192.168.178.xxx:34194,r=app.hydrawise.com/44.236.19.36:443,OPEN,fill=-,flush=-,to=1116380/0}

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:301) ~[?:?]

Caused by: java.io.EOFException: HttpConnectionOverHTTP@1066aff::DecryptedEndPoint@df1913{l=/192.168.178.xxx:34194,r=app.hydrawise.com/44.236.19.36:443,OPEN,fill=-,flush=-,to=1116380/0}

2022-07-21 18:30:41.217 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'hydrawise:account:f4b9xxxf80' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): java.io.IOException: Exception in oauth communication, grant type refresh_token

2022-07-21 18:30:41.224 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'hydrawise:controller:f4b9xxxf80:83xxx9' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)

2022-07-21 18:30:41.174 [DEBUG] [l.api.graphql.HydrawiseGraphQLClient] - Could not send request

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:285) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLQuery(HydrawiseGraphQLClient.java:256) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.queryControllers(HydrawiseGraphQLClient.java:117) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:188) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:206) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:183) ~[?:?]

2022-07-21 18:30:41.201 [DEBUG] [rnal.handler.HydrawiseAccountHandler] - Will try again during next poll period

org.openhab.binding.hydrawise.internal.api.HydrawiseConnectionException: java.io.IOException: Exception in oauth communication, grant type refresh_token

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:307) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLQuery(HydrawiseGraphQLClient.java:256) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.queryControllers(HydrawiseGraphQLClient.java:117) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:188) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:206) ~[?:?]

	at org.openhab.binding.hydrawise.internal.handler.HydrawiseAccountHandler.poll(HydrawiseAccountHandler.java:183) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:285) ~[?:?]

	at org.openhab.binding.hydrawise.internal.api.graphql.HydrawiseGraphQLClient.sendGraphQLRequest(HydrawiseGraphQLClient.java:285) ~[?:?]

```

--- a/bundles/org.openhab.binding.hydrawise/README.md
+++ b/bundles/org.openhab.binding.hydrawise/README.md
@@ -172,7 +172,7 @@ Group SprinklerSensors      "Sensors"       (Sprinkler)
 Group SprinkerForecast      "Forecast"      (Sprinkler)
 
 String SprinkerControllerStatus "Status [%s]" (SprinklerController) {channel="hydrawise:controller:myaccount:123456:controller#status"}
-Number SprinkerControllerLastContact "Last Contact [%d]" (SprinklerController) {channel="hydrawise:controller:myaccount:123456:controller#lastContact"}
+Number SprinkerControllerLastContact "Last Contact [%d]" (SprinklerController) {channel="hydrawise:controller:myaccount:123456:controller#lastcontact"}
 
 Switch SprinklerSensor1 "Sprinler Sensor" (SprinklerSensors) {channel="hydrawise:controller:myaccount:123456:sensor1#active"}
 

--- a/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java
+++ b/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java
@@ -66,7 +66,8 @@ public class HydrawiseAccountHandler extends BaseBridgeHandler implements Access
     private static final String CLIENT_SECRET = "zn3CrjglwNV1";
     private static final String CLIENT_ID = "hydrawise_app";
     private static final String SCOPE = "all";
-    private final List<HydrawiseControllerListener> controllerListeners = new ArrayList<HydrawiseControllerListener>();
+    private final List<HydrawiseControllerListener> controllerListeners = Collections
+            .synchronizedList(new ArrayList<HydrawiseControllerListener>());
     private final HydrawiseGraphQLClient apiClient;
     private final OAuthClientService oAuthService;
     private @Nullable ScheduledFuture<?> pollFuture;
@@ -197,9 +198,11 @@ public class HydrawiseAccountHandler extends BaseBridgeHandler implements Access
                 updateStatus(ThingStatus.ONLINE);
             }
             lastData = response.data.me;
-            controllerListeners.forEach(listener -> {
-                listener.onData(response.data.me.controllers);
-            });
+            synchronized (controllerListeners) {
+                controllerListeners.forEach(listener -> {
+                    listener.onData(response.data.me.controllers);
+                });
+            }
         } catch (HydrawiseConnectionException e) {
             if (retry) {
                 logger.debug("Retrying failed poll", e);

--- a/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java
+++ b/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java
@@ -117,7 +117,9 @@ public class HydrawiseAccountHandler extends BaseBridgeHandler implements Access
     }
 
     public void removeControllerListeners(HydrawiseControllerListener listener) {
-        this.controllerListeners.remove(listener);
+        synchronized (controllerListeners) {
+            this.controllerListeners.remove(listener);
+        }
     }
 
     public @Nullable HydrawiseGraphQLClient graphQLClient() {

--- a/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java
+++ b/bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java
@@ -96,17 +96,26 @@ public class HydrawiseControllerHandler extends BaseThingHandler implements Hydr
     public void initialize() {
         HydrawiseControllerConfiguration config = getConfigAs(HydrawiseControllerConfiguration.class);
         controllerId = config.controllerId;
-        Bridge bridge = getBridge();
-        if (bridge != null) {
-            HydrawiseAccountHandler handler = (HydrawiseAccountHandler) bridge.getHandler();
-            if (handler != null) {
-                handler.addControllerListeners(this);
+        HydrawiseAccountHandler handler = getAccountHandler();
+        if (handler != null) {
+            handler.addControllerListeners(this);
+            Bridge bridge = getBridge();
+            if (bridge != null) {
                 if (bridge.getStatus() == ThingStatus.ONLINE) {
                     updateStatus(ThingStatus.ONLINE);
                 } else {
                     updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
                 }
             }
+        }
+    }
+
+    @Override
+    public void dispose() {
+        logger.debug("Controller Handler disposed.");
+        HydrawiseAccountHandler handler = getAccountHandler();
+        if (handler != null) {
+            handler.removeControllerListeners(this);
         }
     }
 

--- a/bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml
+++ b/bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml
@@ -68,13 +68,13 @@
 		<channels>
 			<channel id="name" typeId="name"/>
 			<channel id="summary" typeId="summary"/>
-			<channel id="lastcontacttime" typeId="lastcontacttime"/>
+			<channel id="lastcontact" typeId="lastcontact"/>
 		</channels>
 	</channel-group-type>
 
 	<!-- Controller -->
 
-	<channel-type id="lastcontacttime" advanced="true">
+	<channel-type id="lastcontact" advanced="true">
 		<item-type>DateTime</item-type>
 		<label>Last Contact Time</label>
 		<description>Last contact time of a controller</description>

